tests/loggen/logline_generator.c

```diff
@@ -110,16 +113,31 @@ generate_log_line(char *buffer, int buffer_length, int syslog_proto, int thread_
 
   /* create time stamps */
   struct timeval now;
-  gettimeofday(&now, NULL);
+  static struct timeval ts_formatted = {0};
   struct tm tm;
-  char stamp[32];
-  localtime_r(&now.tv_sec, &tm);
-  int len = strftime(stamp, sizeof(stamp), "%Y-%m-%dT%H:%M:%S", &tm);
-  memcpy(&buffer[pos_timestamp2], stamp, len);
+  static char stamp[32] = {0};
+  static int len;
 
-  if (syslog_proto)
-    format_timezone_offset_with_colon(stamp, sizeof(stamp), &tm);
+  int check_seq = rate / 10;
+  if (check_seq > 1000)
+    check_seq = 1000;
+  if (check_seq <= 1 || (seq % check_seq) == 0)
+    {
+      g_mutex_lock(&generator_lock);
+      gettimeofday(&now, NULL);
+      if (now.tv_sec != ts_formatted.tv_sec)
+        {
+          localtime_r(&now.tv_sec, &tm);
+          len = strftime(stamp, sizeof(stamp), "%Y-%m-%dT%H:%M:%S", &tm);
+          ts_formatted = now;
+
+          if (syslog_proto)
+            format_timezone_offset_with_colon(stamp, sizeof(stamp), &tm);
+        }
+      g_mutex_unlock(&generator_lock);
+    }
 
+  memcpy(&buffer[pos_timestamp2], stamp, len);
```

Comment:
I think the `memcpy()` call should be protected as well (read-write race).

Since now this will become expensive, I think caching on the caller side or TLS would be suitable.

Reply:
fixed.